### DISCUSSION TITLE
Allow for Maven Credentials to e Set Using Environment Variables

### DISCUSSION THREAD

**Discussion by honeycomb-cheesecake** [sbt/sbt#6897] [Ideas]
Hi, according to [SBT Publishing](https://www.scala-sbt.org/1.x/docs/Publishing.html#credentials), there are two ways of adding credentials, either from a file, or inline. I propose that we allow a third way of adding credentials, i.e. setting the parameters via Environment Variable(s), where you can add the name, host, username and password as 1/more variables which then get picked up during retrieve and publish. SBT already uses SBT_CREDENTIALS to set the default, so it shouldn't be a stretch to add this.

The why is simple, it would make publishing easier from an automated process. Currently, unless there's a way I don't know about and if there is PLEASE TELL ME, I echo the values to a file which is included in the build config or in .sbt/0.13/.credentials or .sbt/1.0/.credentials. It would be so much easier for them to be picked up and used directly.

Using Environment Variables would be easier than echo and more secure than inline. What do people think?

**Comment by seveneves** [comment]
I have never tried it but I think you can create a simple sbt plugin that sets credentials from environment variables. 

For example 

```scala
object CredPlugin extends AutoPlugin {
  
  override def trigger = allRequirements

  override def buildSettings: Seq[Def.Setting[?]] = Seq(
    credentials += Credentials("my maven repo", "mymaven.host.com", sys.env("SBT_USER"), sys.env("SBT_PASS")),
  )
}
```

This plugin can be added to `project/CredPlugin.scala` of each sbt project or built as sbt plugin and then added via `addSbtPlugin` in `proect/plugins.sbt`.

**Reply by honeycomb-cheesecake** [reply]
Hahaha, you know what, thank you so much, I cannot believe I didn't think of this! Also, taking it a step further, I could probably use `sys.env.getOrElse("ENV_VAR", "")` directly in my build.sbt file for the various elements of the Credentials without a plug-in if I wanted. It's always good to talk and get ideas from others, as this was a blind spot, so thanks again.

I'm actually going to close this as there's absolutely no need.

**Comment by laughedelic** [comment]
Related proposal: 

- https://github.com/sbt/sbt/discussions/6386
